eth-repository/src/main/java/net/consensys/cava/eth/repository/BlockchainRepository.java

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package net.consensys.cava.eth.repository;
+
+import net.consensys.cava.bytes.Bytes;
+import net.consensys.cava.bytes.Bytes32;
+import net.consensys.cava.concurrent.AsyncCompletion;
+import net.consensys.cava.concurrent.AsyncResult;
+import net.consensys.cava.eth.Block;
+import net.consensys.cava.eth.BlockHeader;
+import net.consensys.cava.eth.Hash;
+import net.consensys.cava.kv.KeyValueStore;
+
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+
+/**
+ * Repository housing blockchain information.
+ *
+ * This repository allows storing blocks, block headers and metadata about the blockchain, such as forks and head
+ * information.
+ */
+public final class BlockchainRepository {
+
+  private final KeyValueStore chainMetadata;
+  private final KeyValueStore blockStore;
+  private final KeyValueStore blockHeaderStore;
+  private final BlockchainIndex blockchainIndex;
+
+  /**
+   * Initializes a blockchain repository with metadata, placing it in key-value stores.
+   *
+   * @return a new blockchain repository made from the metadata passed in parameter.
+   */
+  public static AsyncResult<BlockchainRepository> init(
```

Comment:
I'd consider converting this to Kotlin and using suspensions, then wrapping it back to `AsyncResult<>`. That's the pattern used most other places in Cava thus far.